src/main/java/org/jenkinsci/plugins/workflow/log/StreamLogStorage.java

```diff
@@ -0,0 +1,331 @@
+/*
+ * The MIT License
+ *
+ * Copyright 2018 CloudBees, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package org.jenkinsci.plugins.workflow.log;
+
+import com.google.common.base.Charsets;
+import com.google.common.primitives.Bytes;
+import hudson.console.AnnotatedLargeText;
+import hudson.console.ConsoleAnnotationOutputStream;
+import hudson.console.ConsoleAnnotator;
+import hudson.console.LineTransformationOutputStream;
+import hudson.model.BuildListener;
+import hudson.model.StreamBuildListener;
+import hudson.model.TaskListener;
+import hudson.util.ByteArrayOutputStream2;
+import java.io.BufferedInputStream;
+import java.io.BufferedOutputStream;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.io.PrintStream;
+import java.io.UnsupportedEncodingException;
+import java.io.Writer;
+import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import javax.annotation.Nonnull;
+import org.apache.commons.io.IOUtils;
+import org.apache.commons.io.input.CountingInputStream;
+import org.jenkinsci.plugins.workflow.flow.FlowExecutionOwner;
+import org.jenkinsci.plugins.workflow.graph.FlowNode;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.Beta;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+import org.kohsuke.stapler.framework.io.ByteBuffer;
+
+/**
+ * Simple implementation of log storage that intermixes node and general messages in a single log bytestream.
+ */
+@Restricted(Beta.class)
+public abstract class StreamLogStorage implements LogStorage {
```

Comment:
🐛  For the record, this is what *currently* requires reading the entire log in order to fetch log information for *any* FlowNode viewed.  This will likely introduce serious performance problems on masters because users will be opening logs for one or more builds from the UI and logs often run 10 MB+, sometimes 100 MB to 1 GB.

If this behavior problem is fixed, this is viable though -- this can be solved bye storing ranges for logs contained in each FlowNode, in the LogAction added here.